### ISSUE TITLE
FreeBSD clients able to freeze a Tokio+Hyper Linux server

### ISSUE THREAD

**Issue by P-E-Meunier** [tokio-rs/tokio-io#92]
I'm having a really nasty bug with Pijul clients on FreeBSD able to freeze my Tokio+Hyper server nest.pijul.com, by just issuing `pijul clone https://nest.pijul.com/pijul_org/pijul`.

That server has a fixed number of threads, each running one Tokio event loop, on which three servers are running:

- A HTTPS server, which gets frozen when the user tries to clone.
- A HTTP server, which still works fine even when the HTTPS server is frozen.
- An SSH server, which also still works fine even when the HTTPS server is frozen.

Moreover, when this happens, the CPU load is normal.

After weeks of debugging, I've come to the conclusion that maybe one of Tokio and Hyper is guilty for that, and I'm not yet sure which. I can't reproduce in a virtual machine (I've tried VirtualBox), but I don't know which network stack gets used then.

Is it possible that because FreeBSD has a totally different networking stack from Linux', socket control works in different ways?

**Comment by carllerche** [comment]
Thanks for the report. Unfortunately, this isn't actionable on our end. If we are to debug further, we would need a reproduction. Also, FreeBSD is "contributor supported". I personally have no way to test on FreeBSD, so it is up to the community to support that.

In theory, FreeBSD is no different than OS X, which is supported. Both are backed by kqueue.

**Comment by carllerche** [comment]
I'm going to close the issue for now, if more info comes up, I can reopen it.

**Comment by P-E-Meunier** [comment]
Wait, this is not really about FreeBSD, the server is running Linux, and  this is a major stability issue, since clients can freeze the Linux server. These clients happen to run FreeBSD, but this might also work with carefully crafted packets sent from Linux.

**Comment by carllerche** [comment]
Either way, we would need a repro. In a large stack like the one you are describing, I would venture to guess the issue is not with Tokio. A freezing server is most likely going to be caused by either

a) A future spinning and not allowing other futures to execute. This would be evident if, while the server is frozen, it is taking up 100% CPU.

b) A buggy future returning `Async::NotReady` and not notifying the task once it becomes ready.

Hope this helps. I do realize that the future ecosystem requires better debugging tools. I will hopefully have some time to work on this early next year.

**Comment by P-E-Meunier** [comment]
Ok, thanks! The process is not using up all the CPU, so it's probably option (b) in Hyper.

**Comment by tailhook** [comment]
Note: option (b) doesn't freeze the server, it just freezes one future. I.e. if one connection is stalled. You should be able to issue another one, and the new one is fully functional.

**Comment by P-E-Meunier** [comment]
Well, that's really not what is happening, I get about 0.2 CPU usage, the other servers (SSH, HTTP) spawn into the same event loop work perfectly (and fast), only one of them doesn't respond anymore. I'm installing FreeBSD now on a spare machine to test.

**Comment by P-E-Meunier** [comment]
Ok, thanks a lot @carllerche for the indications, I finally nailed it. It was indeed an incorrectly spawned future, which I introduced when adapting existing code to recent changes in hyper.

The reason only one person was able to trigger it is, they had a broken version of OpenSSL, which for some reason would never complete the handshake. My handshake code was of the form:
```
incoming.for_each(|(socket, addr)| do_handshake().and_then(handle.spawn(router)))
```
Instead of being spawned into the event loop as well:
```
incoming.for_each(|(socket, addr)| handle.spawn(do_handshake().and_then(router)))
```
